src/tests/unit/tests/background/injector-controller.test.ts

```diff
@@ -116,6 +118,43 @@ describe('InjectorControllerTest', () => {
         await validator.inspectInjectCallback();
         validator.verifyAll();
     });
+
+    test("inject doesn't occur when injection has failed", async () => {
+        const visualizationData = new VisualizationStoreDataBuilder()
+            .with('injectionFailed', true)
+            .build();
+
+        validator
+            .setupTabStore({ id: tabId })
+            .setupVizStoreGetState(visualizationData)
+            .setupInspectStore({ inspectMode: InspectMode.off })
+            .setupInjectScriptsCall(tabId, 0);
+        validator.buildInjectorController().initialize();
+        await validator.inspectInjectCallback();
+        validator.verifyAll();
+    });
+
+    test('inject sends injection failed message when injection errors', async () => {
```

Comment:
Missing a case for the `attempts > 3` transition that should set `injectionFailed: true`?